### ISSUE TITLE
Pipeline: obtain per-image reproducibility seeds when generating multiple images

### ISSUE THREAD

**Issue by pcuenca** [huggingface/diffusers#208]
People are interested in the following workflow: generate a few images from a prompt, select the one you like and tweak the prompt to steer it towards your goal. There are many examples in Twitter and the Discord bot, see context section below.

I managed to replicate this workflow in a demo Space generating one image at a time, and manually setting the seed before each generation. This is sub-optimal, especially when considering high-load services that need to batch requests from several users in order to run them in parallel.

Given that the pipeline is already designed to receive a `Generator`, I'd like to explore if we can (optionally) return **per-image** seeds as part of the generation process.

**Describe the solution you'd like**
I'd like the following to work:

```Python
ldm = DiffusionPipeline.from_pretrained(MODEL_ID)
rng = torch.manual_seed(1337)
prompts = ["Labrador in the style of Vermeer"] * 6
preds = ldm(prompts, generator=rng, return_seeds=True)
images, seeds = preds["sample"], preds["seed"]

rng.manual_seed(seeds[3])
tweaked_image = ldm("Labrador in the style of Hokusai", generator=rng)
```

The same thing should work when batching prompts from several users.

Perhaps something like this is already possible and I haven't looked enough.

One idea is to iterate [here](https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/latent_diffusion/pipeline_latent_diffusion.py#L55) instead of creating the latents all at once. It would have a performance penalty, but I presume it should be negligible compared to the rest of the process. A drawback of this approach is that we'd need to do it for every pipeline, or try to create a helper.

**Describe alternatives you've considered**
* Do nothing, and use the low-level API instead of the pipeline to achieve this goal (I think it should be possible, I plan to test it later). Implementing it in the pipeline would just make it easier for some users to experiment on their own.
* Iterate to obtain just a single image per generation. This wastes GPU memory and is slower in most cases. But it might not be a problem for users that just want to experiment.

**Additional context**

* [This request](https://huggingface.co/CompVis/stable-diffusion-v-1-3/discussions/9#62fdeb768cdcc5d80bc53c68)
* Discord:
![screenshot_from_2022-08-06_21-42-12](https://user-images.githubusercontent.com/1177582/185362725-dc6cf4bc-ef27-4082-a65e-d51025907bc9.png)
(screenshot credit: @osanseviero)

**Comment by anton-l** [comment]
This would be a cool feature indeed! 
Since `torch.Generator()` operates with `ByteTensor` `get_state()` and `set_state()`, for this we might need to replace the `pipe(generator)` argument with just an integer seed, like we've discussed earlier with @patil-suraj @patrickvonplaten

**Comment by spychiatrist** [comment]
Coming here from the HuggingFace comments.  I've implemented a simple workaround for the non-determinism: sequential seeding (for n_iter) now with sub-seed offset (for n_samples).  One new parameter `--seed_offset` allows you to deterministically shift which sample-subseed is selected.  

Here's an example of deterministically shifting down and right in the matrix (previously impossible):

|  | `--n_samples 3 --n_iter 3 --seed 42 --seed_offset 0` | `--n_samples 2 --n_iter 2 --seed 43 --seed_offset 1` |
|---|---|---|
| |  <img src="https://s3.amazonaws.com/moonup/production/uploads/1660835094474-62fc9831cad078c79728f015.png" width="375" height="375" />  |  <img src="https://s3.amazonaws.com/moonup/production/uploads/1660835094446-62fc9831cad078c79728f015.png" width="250" height="250" />   |  

[See this commit (and priors) on my fork of the main repo](https://github.com/spychiatrist/stable-diffusion-improvements/commit/255b6c14c48a6b868c17b2e7f39616d5adaad601)

[Other relevant commit w/ simple sequential seeding](https://github.com/spychiatrist/stable-diffusion-improvements/commit/2adb1b8df3c3797439a65a377e2c8bab977c2ea2)

**Comment by atarashansky** [comment]
@spychiatrist Not a hugging-face dev, but would love to see a PR for this in the base repo.

**Comment by patrickvonplaten** [comment]
> Coming here from the HuggingFace comments. I've implemented a simple workaround for the non-determinism: sequential seeding (for n_iter) now with sub-seed offset (for n_samples). One new parameter `--seed_offset` allows you to deterministically shift which sample-subseed is selected.
> 
> Here's an example of deterministically shifting down and right in the matrix (previously impossible):
> 
> `--n_samples 3 --n_iter 3 --seed 42 --seed_offset 0`	`--n_samples 2 --n_iter 2 --seed 43 --seed_offset 1`
> <img alt="" width="375" height="375" src="https://camo.githubusercontent.com/8f81e038f190452f3601d4fe33939f50a0e534be47f480b15cb8f259d1a872e6/68747470733a2f2f73332e616d617a6f6e6177732e636f6d2f6d6f6f6e75702f70726f64756374696f6e2f75706c6f6164732f313636303833353039343437342d3632666339383331636164303738633739373238663031352e706e67">	<img alt="" width="250" height="250" src="https://camo.githubusercontent.com/45b14241cc69e36e4767d818201c245e80841153189c8b73ed14c24f9ce0f224/68747470733a2f2f73332e616d617a6f6e6177732e636f6d2f6d6f6f6e75702f70726f64756374696f6e2f75706c6f6164732f313636303833353039343434362d3632666339383331636164303738633739373238663031352e706e67">
> [See this commit (and priors) on my fork of the main repo](https://github.com/spychiatrist/stable-diffusion-improvements/commit/255b6c14c48a6b868c17b2e7f39616d5adaad601)
> 
> [Other relevant commit w/ simple sequential seeding](https://github.com/spychiatrist/stable-diffusion-improvements/commit/2adb1b8df3c3797439a65a377e2c8bab977c2ea2)

That's a smart idea with `seed_offset` @spychiatrist ! 
I see the need and the problem, also in favor of finding a good solution here! 

To narrow down the problem here a bit more:

1.
If I take a seed=0 (or a generator state, IMO it doesn't really matter) and I generate gaussian noise of size (4,3), then there is IMO no way to generate gaussian noise of size (1,3) afterwards that corresponds to the first batch_idx of (4,3) except ones generates the tensor of the previous batch size which is API-wise not feasible.

2.
I don't really see how this can be solved if we continue to create images via: https://github.com/huggingface/diffusers/blob/e30e1b89d0a85073f8f5338b86445bf8ed892086/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py#L85
(e.g. generating random noise for the whole batch)
=> we're closer to a solution if we do something like:
```python
        latents = torch.cat([torch.randn(
            (1, self.unet.in_channels, height // 8, width // 8),
            generator=generator,
            device=torch_device,
        ) for _ in range(batch_size)], dim=0)
```
I think, because than we always create the same tensor sizes at least.
In that case, it would also indeed be nicer if we would use seeds (maybe with offsets) or we let the user pass seeds as suggested by @anton-l 

3.
One other way that would put a higher burden on the user, but would very much help us in terms of API and managing complexity would be to maybe just introduce a `latents: Optional[torch.Tensor] = None` argument to the `__call__` function so that the user can pass the noise themselves. This way no random seed or generator is needed at all because the pipeline is deterministic (think that's my favorite approach for now). 

4. 
5. Adding `seed` in addition to `generator` I think is a bit weird. We could replace `generator` with `seed` though if we notice `seed` has more advantages! Also cc @LysandreJik and @sgugger regarding `seed` vs. `generator` in PyTorch since we talked about it a bit.

**Comment by patrickvonplaten** [comment]
@anton-l @patil-suraj what do you think here? IMO, let's just add `latents` as an argument to the pipeline no?

**Comment by pcuenca** [comment]
> @anton-l @patil-suraj what do you think here? IMO, let's just add `latents` as an argument to the pipeline no?

I'm exploring something else in a custom pipeline: pass an array of seeds that defaults to `None`. I can submit a PR for discussion.

**Comment by leszekhanusz** [comment]
So, using the same seed using either diffusers or directly the stable-diffusion library will generate different pictures?

**Comment by pcuenca** [comment]
Fixed in #247.
